### ISSUE TITLE
IndexIVFPQ training OOM with smaller nlist (65536) but works with larger nlist (524288) on 120M x 2560d dataset

### ISSUE THREAD

**Issue by ShacklesLay** [facebookresearch/faiss#4744]
### Summary
I am encountering a counter-intuitive memory issue when training an `IndexIVFPQ` on a large dataset (120M vectors, 2560 dim).

When I set `nlist` to **524,288**, the training proceeds normally.
However, when I reduce `nlist` to **65,536**, the system RAM usage spikes uncontrollably during the pre-processing phase (before GPU utilization goes up), eventually causing the process to be killed.

It seems that reducing `nlist` triggers an aggressive memory allocation strategy during the residual computation or assignment phase that does not occur with a larger `nlist`.

### Environment
*   **FAISS version:** 1.13.0
*   **GPU:** h200
*   **OS:** Linux
*   **RAM:** 1.9T
*   **CUDA:** 12.8

### Reproduction Configuration

**Dataset:**
*   Size: 120,000,000 vectors
*   Dimension: 2560
*   Data type: Float32
*   Total raw data size: ~1.2 TB

**Index String:**
*   `IVF65536,PQ64` (Fails)
*   `IVF524288,PQ64` (Works)

**Code Snippet (Conceptual):**

```python
import faiss
import numpy as np

# Assuming 'training_data' is a memory-mapped file or loaded array of shape (120M, 2560)
# d = 2560

# Scenario 1: Works fine
# nlist = 524288 

# Scenario 2: OOM / Infinite Memory Growth
nlist = 65536 

quantizer = faiss.IndexFlatL2(d)
index = faiss.IndexIVFPQ(quantizer, d, nlist, 64, 8)

# I tried to limit the points per centroid to fix the issue, but it had no effect.
index.cp.max_points_per_centroid = 2048 
index.pq.cp.max_points_per_centroid = 2048

# Move to GPU (Standard GPU resources object)
res = faiss.StandardGpuResources()
gpu_index = faiss.index_cpu_to_gpu(res, 0, index)

# Training
print("Start training...")
# The crash happens here for nlist=65536. 
# Memory grows indefinitely, GPU util is 0%.
gpu_index.train(training_data) 
```

### Observed Behavior

1.  **With `nlist = 524288`:**
    *   System RAM loads the 1.2TB dataset.
    *   Memory increases by ~0.05TB (overhead).
    *   GPU starts running (utilization goes up).
    *   Training completes successfully.

2.  **With `nlist = 65536`:**
    *   System RAM loads the dataset.
    *   **Issue:** RAM usage starts growing beyond the initial 1.2TB.
    *   Unlike the 524k case, it does not stop at a small overhead. It continues to consume all available swap/RAM until the OS kills the process.
    *   **GPU State:** Memory is allocated on the GPU, but **GPU-Util remains at 0%** throughout this growth period. This suggests the bottleneck is in the CPU-side preparation (likely residual computation or assignment) before offloading the PQ training tasks.

### Attempts to Fix
I attempted to set `max_points_per_centroid` to `2048` and `10000` on the CPU index before moving to GPU, assuming it might be related to clustering sample size. This did not change the behavior.

### Questions
1.  Why does a *smaller* `nlist` cause significantly higher peak memory usage during the training phase?
2.  Is FAISS attempting to materialize a massive temporary buffer (e.g., for residuals) that scales inversely with `nlist` or behaves differently when clusters are larger?
3.  Is there a specific parameter to control the batch size of the residual computation step during `IndexIVFPQ` training?

**Comment by ShacklesLay** [comment]
I have identified the root cause and found a solution. I am closing this issue, but documenting the findings below as they may help others encountering similar memory spikes.

### Root Cause Analysis

The issue stems from how Faiss decides whether to **subsample** training data based on `nlist` and `max_points_per_centroid` (default 256), combined with a specific behavior of `index_cpu_to_gpu`.

**1. The Subsampling Logic**
Faiss calculates a threshold to decide if it should subsample the data for clustering:
$$\text{Threshold} = \text{nlist} \times \text{max-points-per-centroid}$$

If `num_train_data > Threshold`, Faiss allocates **new memory** to store the subsampled data.
If `num_train_data <= Threshold`, Faiss uses the **original data pointer** directly (no extra allocation).

**2. Why `nlist=524288` worked:**
$$524,288 \times 256 \approx 134 \text{ million}$$
My data (120M) < Threshold (134M).
**Result:** No subsampling triggered. Faiss used the existing memory pointer. RAM usage remained stable.

**3. Why `nlist=65536` failed (OOM):**
$$65,536 \times 256 \approx 16.7 \text{ million}$$
My data (120M) >> Threshold (16.7M).
**Result:** Subsampling triggered. 
Although the subsampled data (~16.7M vectors) only requires roughly **167 GB**, the process still caused an OOM crash despite having ~700 GB of free RAM. 
This is likely due to **memory fragmentation** (allocating a large contiguous block when 1.2TB is already in use is difficult) and the **additional overhead** of temporary buffers (e.g., permutation arrays for shuffling, pinned memory for GPU transfer staging) required during the copy process.

### The "Gotcha": `index_cpu_to_gpu` Parameter Reset

The reason my initial fix (`cpu_index.cp.max_points_per_centroid = 2048`) failed is that **`faiss.index_cpu_to_gpu` does not copy the `ClusteringParameters` (cp) from the CPU index to the GPU index.**

When the index was moved to the GPU, `max_points_per_centroid` silently reverted to the default `256`, re-triggering the fatal subsampling logic described above.

### Solution

The parameters must be set **on the GPU index object** after the transfer.

```python
# ... setup code ...

# 1. Convert to GPU index
gpu_index = faiss.index_cpu_to_gpu(res, 0, cpu_index)

# 2. CRITICAL FIX: Set parameters on the GPU index instance
# Increase threshold so 120M < nlist * max_points
# For nlist=65536, setting this to 2048 yields a threshold of ~134M
gpu_index.cp.max_points_per_centroid = 2048 

# 3. Train
gpu_index.train(training_data)
